bin/cron/hoc_student_name_cleanup

```diff
@@ -0,0 +1,196 @@
+  end
+
+  opts.on('--sleep=100',
+    Integer,
+    'Amount of time to sleep in milliseconds to sleep between batches. Defaults to ' \
+    '100 msecs if not specified.'
+  ) do |sleep_msecs|
+    options[:sleep_msecs] = sleep_msecs
+  end
+
+  opts.on('--dry-run',
+    'Enables read-only mode where no changes are written to the database'
+  ) do |dry_run|
+    options[:dry_run] = dry_run
+  end
+
+  opts.on('--log-sql',
+    'Logs all sql statements sent to the database'
+  ) do |log_sql|
+    options[:log_sql] = log_sql
+  end
+
+  opts.on('-h', '--help', 'Prints this help message') do
+    puts opts
+    exit
+  end
+end.parse!
+CDO.log.info "Initial parsed options: #{options}"
+
+PEGASUS_DB.loggers << CDO.log if options[:log_sql]
+
+# Get the last row in the db for later use. If there is no last row (i.e. there
+# are no rows in the table), we can exit right away.
+last_row = HOC_ACTIVITY.order(:id).last
+if last_row.nil?
+  CDO.log.info "pegasus.hoc_activity table has no rows, exiting"
+  return
+end
+
+# Calculate the start of the retention window. Rows with a timestamp earlier
+# than this are candidates for cleanup.
+options[:retain_after_date] = Date.today - options[:retention_days]
+
+# If start_id was not specified, read the value from the db
+if options[:start_id].nil?
+  options[:start_id] = Properties.get(START_ID_PROPERTY_KEY) || 0
+end
+
+# If end_id was not specified, determine its value by looking for the first row
+# that is definitely inside the retention window. In the rare case where there
+# are no rows in the retention window, set end_id to 1 past the last row to
+# check all rows
+if options[:end_id].nil?
+  options[:end_id] = HOC_ACTIVITY.select(:id).
+    where {finished_at > options[:retain_after_date]}.
+    order(:finished_at, :id).
+    get(:id) || last_row[:id] + 1
+end
+
+# Options should not change from here on
+options.freeze
+CDO.log.info "Final calculated options: #{options}"
+
+# Main loop - read rows with names one batch at a time and null out entries
+# that are outside the retention window. It is important to keep the batch size
+# small because the name column is not indexed.
+batch_start_id = options[:start_id]
+first_retained_row = nil
+batch_count = 0
+while batch_start_id < options[:end_id]
+  ids_to_cleanup = []
+  last_row_id = nil
+
+  HOC_ACTIVITY.select(:id, :started_at, :finished_at, :pixel_started_at, :pixel_finished_at).
+    where(Sequel.lit('id >= ? AND id < ? AND name IS NOT NULL', batch_start_id, options[:end_id])).
+    order(:id).
+    first(options[:batch_limit]).
+    each do |row|
+      # A row needs to be cleaned up if it has a timestamp that is earlier than the
+      # retention window. Each row has four timestamps, we choose the first non-null
+      # value using the following order: finished_at, pixel_finished_at, started_at,
+      # pixel_started_at. In the rare case where all four timestamps are null, we
+      # assume that the row _should_ be cleaned up.
```

Comment:
How often does this happen? As in, is it possible to have no timestamp but a filled in name?

Reply:
Great question!  As far as I can tell, this _never_ happens but I can't verify this definitively because (1) the code has changed a lot over the years, and (2) the table is too large to query.  I added this check just to be on the safe side so we don't miss any names.

Reply:
The default of clearing the name (in this rare case) sounds good to me. Thanks:)